### ISSUE TITLE
Incorrectly inferred type parameter when candidate types are a subset of each other

### ISSUE THREAD

**Issue by fenduru** [Microsoft/TypeScript#24590]
Apologies for the confusing title, I'm really not sure how to describe this issue succinctly. Feel free to change the title to something more clear.

**TypeScript Version:**  3.0.0-dev.20180601

**Search Terms:** 

Incorrect type inference, widened types, candidate types, subset

**Code**

```ts
let foo: string = 'foo';

// Use a function for indirection to prevent control flow analysis from
// ignoring explicit type annotation and treating bar as number rather than string | number
function getBar(): string | number {
    return 5;
}

let bar: string | number = getBar();

class Container<T> {
    public saved: T;

    constructor(arg: T) {
        this.saved = arg;
    }
}

function assign<T>(x: Container<T>, y: T) {
    x.saved = y;
}

const myContainer = new Container(foo);

assign(myContainer, bar);

const result = myContainer.saved.toUpperCase(); // TypeErrors at runtime
```

**Expected behavior:**

The `assign(myContainer, bar)` should not typecheck as it is attempting to use two incompatible types for the type param `T`. `string | number` is not a common type of `string`. When one candidate type is a subset of the other, the narrowest type should be chosen for inference (leading to an error in the argument position of the wider type)

**Actual behavior:**

`assign<T>` is inferring `string | number` as the type of T even though that is incompatible with the type `string` that the `Container` is parameterized on. This allows the function to assign a non-string value to a string member variable, leading to runtime type errors.

[**Playground Link**](https://www.typescriptlang.org/play/index.html#src=let%20foo%3A%20string%20%3D%20'foo'%3B%0D%0A%0D%0A%2F%2F%20Use%20a%20function%20for%20indirection%20to%20prevent%20control%20flow%20analysis%20from%0D%0A%2F%2F%20ignoring%20explicit%20type%20annotation%20and%20treating%20bar%20as%20number%20rather%20than%20string%20%7C%20number%0D%0Afunction%20getBar()%3A%20string%20%7C%20number%20%7B%0D%0A%20%20%20%20return%205%3B%0D%0A%7D%0D%0A%0D%0Alet%20bar%3A%20string%20%7C%20number%20%3D%20getBar()%3B%0D%0A%0D%0Aclass%20Container%3CT%3E%20%7B%0D%0A%20%20%20%20public%20saved%3A%20T%3B%0D%0A%0D%0A%20%20%20%20constructor(arg%3A%20T)%20%7B%0D%0A%20%20%20%20%20%20%20%20this.saved%20%3D%20arg%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20assign%3CT%3E(x%3A%20Container%3CT%3E%2C%20y%3A%20T)%20%7B%0D%0A%20%20%20%20x.saved%20%3D%20y%3B%0D%0A%7D%0D%0A%0D%0Aconst%20myContainer%20%3D%20new%20Container(foo)%3B%0D%0A%0D%0Aassign(myContainer%2C%20bar)%3B%0D%0A%0D%0Aconst%20result%20%3D%20myContainer.saved.toUpperCase()%3B%20%2F%2F%20TypeErrors%20at%20runtime)

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
See #14829

**Comment by fenduru** [comment]
@RyanCavanaugh not sure if you're linking that as a duplicate, or just as a related topic.

The issue here isn't an inability to express something, rather that the type inference itself is unsound. I think #14829 could be useful on its own for being more expressive/having more control over inference, but in this example I would at minimum expect a compiler error. It also implies an "opt in for safety" which is generally the opposite of what is desired in a typed language.

**Comment by RyanCavanaugh** [comment]
> The issue here isn't an inability to express something, rather that the type inference itself is unsound. 

TS *in general* is unsound against writes to supertyped aliases of objects; this is an intentional design trade-off.

**Comment by fenduru** [comment]
@RyanCavanaugh I do not follow. If the candidate types were `{ string, number }`, the checker would correctly complain that "there is no common type between `string` and `number`". However when the candidate types are `{ string, string | number }` it is inferring the "common type" as `string | number`... but `string | number` *is not* a common type of `string`. If it was a common type I'd be able to do the following:

```ts
let foo: string | number = 5;
let bar: string = foo;
```

**Comment by RyanCavanaugh** [comment]
What you're proposing is that we instead be unsound in this example:
```ts
declare const s: string;
declare const sn: string | number;

declare function pickOneAtRandom<T>(x1: T, x2: T): T;

// p: string, but could be number
const p = pickOneAtRandom(s, sn);
```

**Comment by fenduru** [comment]
@RyanCavanaugh that example would not be unsound, it would type error because it would choose `string` as `T`, and in the `x2` position you'd be trying to pass in a `string | number` to a `string` argument.

**Comment by fenduru** [comment]
Your `pickOneAtRandom` is not truthful though by saying that the function is parameterized on one type - it is not, it is parameterized on two types, and should be written like:

```ts
function pickOneAtRandom<T, U>(x1: T, x2: U): T | U
```

If the above is desired though, then how come the following does not infer T to be the union of the types?

```ts
declare const s: string;
declare const sn: number;

declare function pickOneAtRandom<T>(x1: T, x2: T): T;

// p: string, but could be number
const p = pickOneAtRandom(s, sn);
```

The rules around type inference are inconsistent here - sometimes the inference algorithm does a union of types, and sometimes an intersection of types.

**Comment by fenduru** [comment]
I played around with the `NoInfer<T>` type suggested in the other thread a bit more and I think I was able to get it doing what I need in my use case. I thought this wasn't working because I was testing in the playground - and there seems to be a difference between `const foo: GraphNode<number> = 'foo'` (doesn't error like it should) vs. `declare foo: GraphNode<number>` (can also be achieved by importing from another file, or having a function that returns that type. this seems like it might be a bug as well - I'll open another issue.)
